data_safe_haven/infrastructure/stack_manager.py

```diff
@@ -1,24 +1,113 @@
 """Deploy with Pulumi"""
 import logging
-import os
 import pathlib
 import shutil
 import subprocess
 import time
 from contextlib import suppress
 from importlib import metadata
+from shutil import which
 from typing import Any
 
 from pulumi import automation
+import typer
 
 from data_safe_haven.config import Config
 from data_safe_haven.exceptions import DataSafeHavenAzureError, DataSafeHavenPulumiError
-from data_safe_haven.external import AzureApi, AzureCli
+from data_safe_haven.external import AzureApi
 from data_safe_haven.functions import replace_separators
 from data_safe_haven.infrastructure.stacks import DeclarativeSHM, DeclarativeSRE
 from data_safe_haven.utility import LoggingSingleton
 
 
+def handle_pulumi_login(config) -> None:
+    account = PulumiAccount(config)
+
+    if not account.confirm():
+        account.login()
+        if not account.confirm():
+            raise typer.Exit()
```

Comment:
Is it better to raise one of our own exceptions here or does `typer.Exit()` give a cleaner shutdown?

Reply:
At the moment it gives a 0 return code (which feels a bit odd as it needs to be raised).
It might be more sensible to give a non-zero return code here though.